File: backend/repositories/scheduleRepository.js
import db from '../config/db.js';

/**
 * Get schedules for a user within a date range.
 * @param {number} userId
 * @param {string} startDate (YYYY-MM-DD)
 * @param {string} endDate (YYYY-MM-DD)
 * @returns {Promise<Array>}
 */
export const getSchedulesByRange = async (userId, startDate, endDate) => {
  const query = `
        SELECT us.*, s.name as shift_name, s.start_time, s.end_time, s.flexible_minutes
        FROM user_schedules us
        JOIN shifts s ON us.shift_id = s.id
        WHERE us.user_id = ? AND us.date BETWEEN ? AND ?
        ORDER BY us.date ASC
    `;
  const [rows] = await db.query(query, [userId, startDate, endDate]);
  return rows;
};

/**
 * Get a single schedule for a user on a specific date.
 * @param {number} userId
 * @param {string} date (YYYY-MM-DD)
 * @returns {Promise<Object|null>}
 */
export const getScheduleByDate = async (userId, date) => {
  const query = `
        SELECT us.*, s.name as shift_name, s.start_time, s.end_time, s.flexible_minutes, s.work_days
        FROM user_schedules us
        JOIN shifts s ON us.shift_id = s.id
        WHERE us.user_id = ? AND us.date = ?
    `;
  const [rows] = await db.query(query, [userId, date]);
  return rows[0] || null;
};

/**
 * Create or Update (Upsert) a schedule.
 * @param {Object} scheduleData { userId, shiftId, date, createdBy }
 */
export const upsertSchedule = async ({ userId, shiftId, date, createdBy }) => {
  const query = `
        INSERT INTO user_schedules (user_id, shift_id, date, created_by)
        VALUES (?, ?, ?, ?)
        ON DUPLICATE KEY UPDATE shift_id = VALUES(shift_id), created_by = VALUES(created_by)
    `;
  const [result] = await db.query(query, [userId, shiftId, date, createdBy]);
  return result;
};

/**
 * Delete a schedule.
 * @param {number} userId
 * @param {string} date
 */
export const deleteSchedule = async (userId, date) => {
  const query = `DELETE FROM user_schedules WHERE user_id = ? AND date = ?`;
  const [result] = await db.query(query, [userId, date]);
  return result;
};

/**
 * Bulk create schedules can be implemented if needed,
 * but looping upsert is simpler for now unless perf helps.
 */

File: backend/internal/repository/schedule_repository.go
package repository

import (
	"context"

	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/jmoiron/sqlx"
)

type ScheduleRepository interface {
	GetByRange(ctx context.Context, userID int, startDate string, endDate string) ([]dto.ScheduleResponse, error)
	GetByDate(ctx context.Context, userID int, date string) (*dto.ScheduleResponse, error)
	Upsert(ctx context.Context, userID int, shiftID int, date string, createdBy *int) error
	Delete(ctx context.Context, userID int, date string) error
}

type scheduleRepositoryImpl struct {
	db *sqlx.DB
}

func NewScheduleRepository(db *sqlx.DB) ScheduleRepository {
	return &scheduleRepositoryImpl{db: db}
}

func (r *scheduleRepositoryImpl) GetByRange(ctx context.Context, userID int, startDate string, endDate string) ([]dto.ScheduleResponse, error) {
	var schedules []dto.ScheduleResponse
	query := `
		SELECT us.id, us.user_id, us.shift_id, us.date, 
		       s.name as shift_name, s.start_time, s.end_time, s.flexible_minutes
		FROM user_schedules us
		JOIN shifts s ON us.shift_id = s.id
		WHERE us.user_id = ? AND us.date BETWEEN ? AND ?
		ORDER BY us.date ASC
	`
	err := r.db.SelectContext(ctx, &schedules, query, userID, startDate, endDate)
	return schedules, err
}

func (r *scheduleRepositoryImpl) GetByDate(ctx context.Context, userID int, date string) (*dto.ScheduleResponse, error) {
	var schedule dto.ScheduleResponse
	query := `
		SELECT us.id, us.user_id, us.shift_id, us.date, 
		       s.name as shift_name, s.start_time, s.end_time, s.flexible_minutes
		FROM user_schedules us
		JOIN shifts s ON us.shift_id = s.id
		WHERE us.user_id = ? AND us.date = ?
	`
	err := r.db.GetContext(ctx, &schedule, query, userID, date)
	if err != nil {
		return nil, err
	}
	return &schedule, nil
}

func (r *scheduleRepositoryImpl) Upsert(ctx context.Context, userID int, shiftID int, date string, createdBy *int) error {
	query := `
		INSERT INTO user_schedules (user_id, shift_id, date, created_by)
		VALUES (?, ?, ?, ?)
		ON DUPLICATE KEY UPDATE shift_id = VALUES(shift_id), created_by = VALUES(created_by)
	`
	_, err := r.db.ExecContext(ctx, query, userID, shiftID, date, createdBy)
	return err
}

func (r *scheduleRepositoryImpl) Delete(ctx context.Context, userID int, date string) error {
	query := `DELETE FROM user_schedules WHERE user_id = ? AND date = ?`
	_, err := r.db.ExecContext(ctx, query, userID, date)
	return err
}
